### PR TITLE
Ensure relay chain parent is the relay chain head 

### PR DESCRIPTION
**Pull request by cecton**
Related to #7

### CODE REVIEW

File: collator/src/lib.rs [hunk 1]
```diff
@@ -276,11 +282,13 @@ where
 				parachain::HeadData(head_data.encode()),
 			);
 
-			wait_to_announce.lock().wait_to_announce(
-				hash,
-				relay_chain_parent,
-				encoded_header,
-			);
+			if relay_chain_parent == best_hash {
```
Comment (bkchr) [inline collator/src/lib.rs:285]: This is not required. This code is only called with the best hash.

File: network/src/lib.rs [hunk 2]
```diff
@@ -106,6 +109,23 @@ where
 			},
 		} = gossip_statement;
 
+		// Check that the relay chain parent of the block is the relay chain head
+		let best_number = self.polkadot_client.info().best_number;
+
+		match self.polkadot_client.number(relay_chain_leaf) {
```
Comment (bkchr) [inline network/src/lib.rs:115]:
```suggestion
		match self.polkadot_client.number(relay_chain_leaf)? {
```

Doesn't this work?

File: network/src/lib.rs [hunk 3]
```diff
@@ -106,6 +109,23 @@ where
...
+				return Err(Box::new(ClientError::Backend(format!(
+					"could not find block number for {}: {}",
+					relay_chain_leaf, err,
+				))));
+			},
+			Ok(Some(x)) if x == best_number => {},
+			Ok(None) => {
+				return Err(Box::new(ClientError::UnknownBlock(relay_chain_leaf.to_string())));
+			},
+			Ok(Some(_)) => return Ok(Validation::Failure),
```
Comment (bkchr) [inline network/src/lib.rs:126]: Having a trace log would be nice.

File: network/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+		"validation should fail if the relay chain leaf does not exist"
+	);
+	assert!(matches!(
+		*res.unwrap().downcast::<ClientError>().unwrap(),
+		ClientError::UnknownBlock(_)
+	));
+}
+
+#[test]
+fn check_relay_parent_does_not_fail() {
```
Comment (bkchr) [inline network/src/tests.rs:161]: It does not fail but the implementation checks for failing?

File: network/src/tests.rs [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+	) -> std::result::Result<sc_client_api::blockchain::BlockStatus, sp_blockchain::Error> {
+		Ok(sc_client_api::blockchain::BlockStatus::Unknown)
+	}
+
+	fn number(
+		&self,
+		hash: PHash,
+	) -> std::result::Result<Option<NumberFor<PBlock>>, sp_blockchain::Error> {
+		Ok(match hash.as_fixed_bytes() {
+			&[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] => {
```
Comment (bkchr) [inline network/src/tests.rs:436]: We should use `if hash == H256::from_low_u64_be(1)` etc here. That is much shorter.

File: network/src/lib.rs (hunk 2 above)
Comment (cecton) [reply network/src/lib.rs:115]: hmmm no :disappointed: 

```
   --> network/src/lib.rs:115:54
    |
115 |         match self.polkadot_client.number(relay_chain_leaf)? {
    |                                                            ^ the trait `std::convert::From<sp_blockchain::error::Error>` is not implemented for `std::boxed::Box<dyn std::error::Error + std::marker::Send>`
    |
```

File: network/src/tests.rs (hunk 4 above)
Comment (cecton) [reply network/src/tests.rs:161]: yeah makes no sense, will rename that

File: network/src/tests.rs (hunk 5 above)
Comment (cecton) [reply network/src/tests.rs]: done

File: network/src/lib.rs (hunk 3 above)
Comment (cecton) [reply network/src/lib.rs]: done

**Pull request merged**